### PR TITLE
[SPARK-19893][SQL] should not run DataFrame set oprations with map type

### PR DESCRIPTION
**Pull request by cloud-fan**
## What changes were proposed in this pull request?

In spark SQL, map type can't be used in equality test/comparison, and `Intersect`/`Except`/`Distinct` do need equality test for all columns, we should not allow map type in `Intersect`/`Except`/`Distinct`.

## How was this patch tested?

new regression test

### CODE REVIEW

**Comment by cloud-fan** [comment]
cc @yhuai @sameeragarwal @hvanhovell

**Comment by cloud-fan** [comment]
retest this please

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/CheckAnalysis.scala [hunk 1]
```diff
@@ -319,7 +322,7 @@ trait CheckAnalysis extends PredicateHelper {
               // Check if the data types match.
               dataTypes(child).zip(ref).zipWithIndex.foreach { case ((dt1, dt2), ci) =>
                 // SPARK-18058: we shall not care about the nullability of columns
-                if (TypeCoercion.findWiderTypeForTwo(dt1.asNullable, dt2.asNullable).isEmpty) {
+                if (!dt1.sameType(dt2)) {
```
Comment (hvanhovell) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/CheckAnalysis.scala]: We shouldn't change this. This makes sure we generate the correct error message for union. The problem is that an earlier pair of columns might not be the same but castable, while a later pair column is not the same and we can't cast them; the error should show the latter pair of columns and not the first one (which happens if we revert this).

Comment (hvanhovell) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/CheckAnalysis.scala]: See #16882 for more information.

**Comment by hvanhovell** [comment]
@cloud-fan can we also check `Distinct`? We rewrite this into an `Aggregate` in the optimizer, which has the same problem with `MapType`.

**Comment by hvanhovell** [comment]
LGTM pending jenkins

File: sql/core/src/test/scala/org/apache/spark/sql/DataFrameSuite.scala [hunk 2]
```diff
@@ -1703,4 +1703,17 @@ class DataFrameSuite extends QueryTest with SharedSQLContext {
...
+
+  test("SPARK-19893: cannot run intersect/except/distinct with map type") {
+    val df = spark.range(1).select(map(lit("key"), $"id"))
+    val e = intercept[AnalysisException](df.intersect(df))
+    assert(e.message.contains(
+      "Cannot have map type columns in DataFrame which calls intersect/except/distinct"))
+    val e2 = intercept[AnalysisException](df.except(df))
+    assert(e2.message.contains(
+      "Cannot have map type columns in DataFrame which calls intersect/except/distinct"))
+    val e3 = intercept[AnalysisException](df.distinct())
```
Comment (hvanhovell) [inline sql/core/src/test/scala/org/apache/spark/sql/DataFrameSuite.scala]: For some reason I don't understand, the code path for df.distinct() actually creates an aggregate, whereas the SQL code path uses the `Distinct` operator. So we need to actually issue a sql statement here.

**Comment by cloud-fan** [comment]
thanks for the review, merging to master!

**Comment by cloud-fan** [comment]
backported to 2.1/2.0

**Comment by srowen** [comment]
@cloud-fan I think branch 2.0 fails after this commit unfortunately, such as in
https://amplab.cs.berkeley.edu/jenkins/view/Spark%20QA%20Test%20(Dashboard)/job/spark-branch-2.0-test-maven-hadoop-2.4/990/

```
[info] Compiling 174 Scala sources and 19 Java sources to /home/jenkins/workspace/spark-branch-2.0-test-maven-hadoop-2.4/sql/core/target/scala-2.11/test-classes...
[error] /home/jenkins/workspace/spark-branch-2.0-test-maven-hadoop-2.4/sql/core/src/test/scala/org/apache/spark/sql/DataFrameSuite.scala:1704: Missing closing brace `}' assumed here
[error] }
[error] ^
```

**Comment by cloud-fan** [comment]
I'm pretty sorry about this, I have fixed it in branch-2.0, @srowen thanks for pointing this out!

**Comment by magic890** [comment]
@iRakson @cloud-fan @tarekbecker @w4-sjcho @yhuai @sameeragarwal @hvanhovell 
Any plan to re-enable Intersect / Except / Distinct operations for Map type?
The current behaviour prevent even the `left_anti` join, not only the hash operations.

**Comment by cloud-fan** [comment]
You can turn map into array by `map_entries` function to work around it. We don't plan to re-enable it unless we make map type orderable.

**Pull request closed without merging**